cloud/src/autogluon/cloud/utils/ag_sagemaker.py

```diff
@@ -163,3 +157,95 @@ def transformer(
             assemble_with=assemble_with,
             **kwargs
         )
+
+
+class AutoGluonRepackInferenceModel(AutoGluonSagemakerInferenceModel):
+    """
+        Custom implementation to force repack of inference code into model artifacts
+    """
+
+    def prepare_container_def(
+        self,
+        instance_type=None,
+        accelerator_type=None,
+        serverless_inference_config=None,
+    ):  # pylint: disable=unused-argument
+        deploy_key_prefix = fw_utils.model_code_key_prefix(
+            self.key_prefix, self.name, self.image_uri
+        )
+        deploy_env = copy.deepcopy(self.env)
+        self._upload_code(deploy_key_prefix, repack=True)
+        deploy_env.update(self._script_mode_env_vars())
+        return sagemaker.container_def(
+            self.image_uri,
+            self.repacked_model_data or self.model_data,
+            deploy_env,
+            image_config=self.image_config,
+        )
+
+
+class AutoGluonNonRepackInferenceModel(AutoGluonSagemakerInferenceModel):
+    """
+    Custom implementation to force no repack of inference code into model artifacts.
+    This requires inference code already present in the trained artifacts, which is created during CloudPredictor training.
+    """
+
+    def prepare_container_def(
+        self,
+        instance_type=None,
+        accelerator_type=None,
+        serverless_inference_config=None,
+    ):  # pylint: disable=unused-argument
+        deploy_env = copy.deepcopy(self.env)
+        deploy_env.update(self._script_mode_env_vars())
+        deploy_env[SCRIPT_PARAM_NAME.upper()] = os.path.basename(deploy_env[SCRIPT_PARAM_NAME.upper()])
+        deploy_env[DIR_PARAM_NAME.upper()] = '/opt/ml/model/code'
```

Comment:
nit: this can be extracted into a separate method, then the rest of the code can be moved into abstract class and reuse - this would make re-pack vs non-repack difference easier to read:
```python
class AbstractAutoGluonInferenceModel(AutoGluonSagemakerInferenceModel, ABC):

    def prepare_container_def(
            self,
            instance_type=None,
            accelerator_type=None,
            serverless_inference_config=None,
    ):  # pylint: disable=unused-argument
        deploy_env = copy.deepcopy(self.env)
        deploy_env.update(self._script_mode_env_vars())
        self._prepare_container_def(deploy_env)

        return sagemaker.container_def(
            self.image_uri,
            self.model_data,
            deploy_env,
            image_config=self.image_config,
        )

    @abstractmethod
    def _prepare_container_def(self, deploy_env):
        raise NotImplementedError


class AutoGluonNonRepackInferenceModel(AbstractAutoGluonInferenceModel):

    def _prepare_container_def(self, deploy_env):
        deploy_env[SCRIPT_PARAM_NAME.upper()] = os.path.basename(deploy_env[SCRIPT_PARAM_NAME.upper()])
        deploy_env[DIR_PARAM_NAME.upper()] = '/opt/ml/model/code'


class AutoGluonRepackInferenceModel(AbstractAutoGluonInferenceModel):

    def _prepare_container_def(self, deploy_env):
        deploy_key_prefix = fw_utils.model_code_key_prefix(
            self.key_prefix, self.name, self.image_uri
        )
        self._upload_code(deploy_key_prefix, repack=True)

```

Reply:
Actually your provided code snippet doesn't work. The thing is `self._upload_code` will affect the result of `self._script_mode_env_vars()` hence affecting the `deploy_env` being updated. If we move the `deploy_env.update(self._script_mode_env_vars())` into `_prepare_container_def `, there's very little thing the two method shares so a `ABC` is likely not needed.

Reply:
This is a side effect, which should be avoided. Can we make the side effect updating vars explicit in this method?